# Linear Programming Practice (Simplex Method)

## Minimize the costs for blending orange juice

First let's read the specs provided from the different juice suppliers, I am using pandas only to visualize the table automatically

In [5]:
import pandas as pd
import numpy as np

df = pd.read_csv('Supplier Specs.csv')

df

Varietal      Region  Qty Available (1,000 Gallons)  \
0           Hamlin      Brazil                            672   
1          Mosambi       India                            400   
2         Valencia     Florida                           1200   
3           Hamlin  California                            168   
4          Gardner     Arizona                             84   
5          Sunstar       Texas                            210   
6         Jincheng       China                            588   
7            Berna       Spain                            168   
8            Verna      Mexico                            300   
9   Biondo Commune       Egypt                            210   
10      Belladonna       Italy                            180   

    Brix / Acid Ratio  Acid (%)  Astringency (1-10 Scale)  Color (1-10 Scale)  \
0                10.5    0.0060                         3                   3   
1                 6.5    0.0140                         7                   1   
2                12.0    0.0095                         3                   3   
3                11.0    0.0100                         3                   5   
4                12.0    0.0070                         1                   5   
5                10.0    0.0070                         1                   5   
6                 9.0    0.0135                         7                   3   
7                15.0    0.0110                         4                   8   
8                 8.0    0.0130                         8                   3   
9                13.0    0.0130                         3                   5   
10               14.0    0.0050                         3                   9   

    Price (per 1K Gallons)  Shipping (per 1K Gallons)  
0                      500                        100  
1                      310                        150  
2                      750                          0  
3                      600                         60  
4                      600                         75  
5                      625                         50  
6                      440                        120  
7                      600                        110  
8                      300                         90  
9                      460                        130  
10                     505                        115

* Brix/Acid ratio: Brix is a measure of sweetness in the juice, so Brix/Acid ratio is a
measure of sweetness to tartness, which in the end, is really what orange juice is
all about.
* Acid (%): Acid as a percentage of the juice is broken out individually, because at
a certain point, it doesn’t really matter how sweet the juice is, it’s still too acidic.
* Astringency (1–10 scale): A measure of the “green” quality of the juice. It’s that
bitter, unripe, planty fl avor that can creep in. This scale is assessed by a panel of
tasters at each juicing facility on a scale of 1–10.

## Our requirements for the optimization are the following:

* Lowest cost purchase plan for January, February, and March that meets a projected demand of 600,000 gallons of juice in January and February and 700,000 gallons in March.
* Our company has entered an agreement with the state of Florida which provides the company tax incentives so long as the company buys at least 40 percent of its juice each month from Florida Valencia growers. Under no circumstances are you to violate this agreement.
* The Brix/Acid ratio (BAR) must stay between 11.5 and 12.5 in each month’s blend.
* The acid level must remain between 0.75 and 1 percent.
* The astringency level must stay at 4 or lower.
* Color must remain between 4.5 and 5.5. Not too watery, not too dark.

## Real quickly shove those requirements into an outline of an LP formulation:
* **Objective**: Minimize procurement costs.
* **Decisions**: Amount of each juice to buy each month
* **Constraints**:
    * Demand
    * Supply
    * Florida Valencia requirement
    * Flavor
    * Color

# Formulating the problem

I will denote the variables which we need to an optimal solution for as $x_0, x_1, ...$. For this problem these are the quantities of juice which we need to buy from each supplier. In our case we have 11 different suppliers which results in 11 variables, named from $x_0$ to $x_{10}$. Let's assume that $x_0$ is the juice from Brazil, $x_1$ is the juice from India and etc.



## Objective function

The aim which we have is to minimize the total cost (purchase + shipping). Well the cost depends solely on the amount of juice which we buy, thus we can write that the total cost for buying 1000 galons of juice from Brazil will be $500 + 100 = 600$. I didn't write 1000 anywhere as the purchase and shipping costs are priveded per 1000 galons.

Having all this in mind we can write the objective function in the following form:

$(500+100)x_0 + (310+150)x_1 + 750x_2 + (600+60)x_3 + (600+75)x_4 + (625+50)x_5 + (440+120)x_6 + (600+110)x_7 + (300+90)x_8 + (460+130)x_9 + (505+115)x_{10}$

And for the sake of clarity once again:

$f = 600x_0 + 460x_1 + 750x_2 + 660x_3 + 675x_4 + 675x_5 + 560x_6 + 710x_7 + 390x_8 + 590x_9 + 620x_{10}$

$$f = \sum{((purchase_i+shipping_i)x_i)}$$

## Demand (equality constraint)

In the problem we also have to account for total amount of juice which we blend:

* Lowest cost purchase plan for January, February, and March that meets a projected demand of 600,000 gallons of juice in January and February and 700,000 gallons in March.

This translates into the following equalities:

$ D = \sum{x_i} = 600 $ - for January and February

$ D = \sum{x_i} = 700 $ - for March

## Supply (bounds)

On the other hand in the supplier table we see that we do not have a limited supply of $x_0, x_1, x_2, ...$ This fact imposes the following requirements on our variables:

$$x_0 \le 672$$
$$x_1 \le 400$$
$$x_2 \le 1200$$
$$x_3 \le 168$$
$$x_4 \le 84$$
$$x_5 \le 210$$
$$x_6 \le 588$$
$$x_7 \le 168$$
$$x_8 \le 300$$
$$x_9 \le 210$$
$$x_{10} \le 180$$

Although it looks good, these limits are not sufficient for simulating a realistic scenario. Imagine that we decide to purchase negative amout of juice form some/all suppliers. In addition in the problem statement we see that each month we need to purchase certain amount of juice depending on our production (40%) so we can keep the negotiated tax benefits. Combining these two points we result with the following limits for our variables:

$$0 \le x_0 \le 672$$
$$0 \le x_1 \le 400$$
$$0.4D \le x_2 \le 1200$$
$$0 \le x_3 \le 168$$
$$0 \le x_4 \le 84$$
$$0 \le x_5 \le 210$$
$$0 \le x_6 \le 588$$
$$0 \le x_7 \le 168$$
$$0 \le x_8 \le 300$$
$$0 \le x_9 \le 210$$
$$0 \le x_{10} \le 180$$

## Flavor and Color (inequality constraints)

The last point which we need to take into account is the quality of the juice which we will blend. Let's start with the first criteria:

* **The Brix/Acid ratio (BAR) must stay between 11.5 and 12.5 in each month’s blend.**

This can be formulated as the weighted avarage between the amount of juice and their BAR to be within the rage $[11.5;12.5]$ or:

$$11.5 \le \frac{\sum{(BAR_i*x_i)}}{\sum{x_i}} \le 12.5$$

The next point is:

* **The acid level must remain between 0.75 and 1 percent.**

Following the logic from the previous point, this can be written as:

$$0.0075 \le \frac{\sum{(Acid_i*x_i)}}{\sum{x_i}} \le 0.01$$

The next point is:

* **The astringency level must stay at 4 or lower**

which can be written as:

$$0 \le \frac{\sum{(Astringency_i*x_i)}}{\sum{x_i}} \le 4.0$$

and the last point is:

* **Color must remain between 4.5 and 5.5. Not too watery, not too dark.**

which can be written as:

$$4.5 \le \frac{\sum{(Color_i*x_i)}}{\sum{x_i}} \le 5.5$$

These inequalities are easy to read but we would have some trouble implementing them in Python directly as we only need the coefficients in front of each $x_i$. Thus we need to rewrite them in the following general form:

$$min \le \frac{\sum{(k_ix_i)}}{\sum{x_i}} \le max$$

$$min\sum{x_i} \le \sum{(k_ix_i)} \le max\sum{x_i}$$

In other words if we want to keep our coefficients as they are given in the supplier table we just need to scale our limits for the equalities with the current total amount of juice which we blend per month.

# Translating this into Python

I am going to use the [PuLP](https://pythonhosted.org/PuLP/index.html) library for performing the optimization.

Ok Let's start!

In [126]:
from pulp import *

# define our demand for January, February and March in 1,000 Gallons
D = {
    'Jan' : 600,
    'Feb' : 600,
    'Mar' : 700
}

supplyCapacity = df.loc[:,'Qty Available (1,000 Gallons)']

def optimize_single_month(df, D, month, supplyCapacity):
    
    # define the optimization problem
    problem = LpProblem(name="Juice Blending Problem",sense=LpMinimize)

    # define the names of the optimization variables
    suppliers = df.loc[:,'Region']
    
    # define the limits for the optimization variables
    suppliers_lower = np.zeros(11)
    suppliers_lower[2] = 0.4*D[month]
    suppliers_upper = supplyCapacity
    
    # define the optimization variables 
    x = []
    
    for i in range(0,len(suppliers)):
        x.append(LpVariable(name=suppliers[i], lowBound=suppliers_lower[i], upBound=suppliers_upper[i], cat='Continuous'))

    # define the constraint coefficients
    costs = df.loc[:,'Price (per 1K Gallons)'] + df.loc[:,'Shipping (per 1K Gallons)']
    bar   = df.loc[:,'Brix / Acid Ratio']
    acid  = df.loc[:,'Acid (%)']
    astringency = df.loc[:,'Astringency (1-10 Scale)']
    color = df.loc[:,'Color (1-10 Scale)']
    
    # adding the objective function to the optimization model (the expression which we want to minimize)
    problem += lpSum([coeff*x for coeff,x in zip(costs,x)]), "total cost of oranges"
    
    # adding the constraints to the optimization model
    problem += lpSum(x) == D[month], 'Demand constraint'
    
    problem += lpSum([coeff*x for coeff,x in zip(bar,x)]) >= 11.5*lpSum(x), 'Lower BAR constraint'
    problem += lpSum([coeff*x for coeff,x in zip(bar,x)]) <= 12.5*lpSum(x), 'Upper BAR constraint'
    
    problem += lpSum([coeff*x for coeff,x in zip(acid,x)]) >= 0.0075*lpSum(x), 'Lower Acid constraint'
    problem += lpSum([coeff*x for coeff,x in zip(acid,x)]) <= 0.0100*lpSum(x), 'Upper Acid constraint'
    
    problem += lpSum([coeff*x for coeff,x in zip(astringency,x)]) >= 0*lpSum(x), 'Lower Astringency constraint'
    problem += lpSum([coeff*x for coeff,x in zip(astringency,x)]) <= 4*lpSum(x), 'Upper Astringency constraint'
    
    problem += lpSum([coeff*x for coeff,x in zip(color,x)]) >= 4.5*lpSum(x), 'Lower Color constraint'
    problem += lpSum([coeff*x for coeff,x in zip(color,x)]) <= 5.5*lpSum(x), 'Upper Color constraint'
       
    # executing the optimization
    problem.solve()
    
    # printing the status of the solution [“Not Solved”, “Infeasible”, “Unbounded”, “Undefined”, “Optimal”]
    print("Solution Status:", LpStatus[problem.status], '\n')
    
    # print the optimal values for each variable x
    print('---------- Optimal Values ----------')
    [print(var.name, '=', var.varValue) for var in problem.variables()]
    
    print('\n---------- Solution ----------')
    print("Objective Function Value = ", value(problem.objective))

optimize_single_month(df, D, 'Jan', supplyCapacity)

Solution Status: Optimal 

---------- Optimal Values ----------
Arizona = 0.0
Brazil = 0.0
California = 0.0
China = 0.0
Egypt = 109.73684
Florida = 240.0
India = 0.0
Italy = 108.15789
Mexico = 126.31579
Spain = 0.0
Texas = 15.789474

---------- Solution ----------
Objective Function Value =  371723.68045


## Intermediate Result

We got the exact result as in excel.

![title](img/Jan.png)

## Optimizing for several months

So far our optimization model has proven itself when applied for a single month but meanwhile the task we need to solve has changed - it appears that the supply **'Qty Available (1,000 Gallons)'** was not given per month but for a period of 3 months, in other words we cannot simply reorder the quantities which we got from the optimization for the next months as we can see from the excel screen shot our supplier from Egypt for example will run out of oranges by March. Thus now we need to optimize the overall purchase cost for period of 3 months and accounting for the quantities as well.

In [185]:
def optimize_single_month(df, D, month, supplyCapacity):
    
    # define the optimization problem
    problem = LpProblem(name="Juice Blending Problem",sense=LpMinimize)

    # define the names of the optimization variables
    suppliers = df.loc[:,'Region']
    
    # define the limits for the optimization variables
    suppliers_lower = np.zeros(11)
    suppliers_lower[2] = 0.4*D[month]
    suppliers_upper = supplyCapacity
    
    # define the optimization variables 
    x = []
    
    for i in range(0,len(suppliers)):
        x.append(LpVariable(name=suppliers[i], lowBound=suppliers_lower[i], upBound=suppliers_upper[i], cat='Continuous'))

    # define the constraint coefficients
    costs = df.loc[:,'Price (per 1K Gallons)'] + df.loc[:,'Shipping (per 1K Gallons)']
    bar   = df.loc[:,'Brix / Acid Ratio']
    acid  = df.loc[:,'Acid (%)']
    astringency = df.loc[:,'Astringency (1-10 Scale)']
    color = df.loc[:,'Color (1-10 Scale)']
    
    # adding the objective function to the optimization model (the expression which we want to minimize)
    problem += lpSum([coeff*x for coeff,x in zip(costs,x)]), "total cost of oranges"
    
    # adding the constraints to the optimization model
    problem += lpSum(x) == D[month], 'Demand constraint'
    
    problem += lpSum([coeff*x for coeff,x in zip(bar,x)]) >= 11.5*lpSum(x), 'Lower BAR constraint'
    problem += lpSum([coeff*x for coeff,x in zip(bar,x)]) <= 12.5*lpSum(x), 'Upper BAR constraint'
    
    problem += lpSum([coeff*x for coeff,x in zip(acid,x)]) >= 0.0075*lpSum(x), 'Lower Acid constraint'
    problem += lpSum([coeff*x for coeff,x in zip(acid,x)]) <= 0.0100*lpSum(x), 'Upper Acid constraint'
    
    problem += lpSum([coeff*x for coeff,x in zip(astringency,x)]) >= 0*lpSum(x), 'Lower Astringency constraint'
    problem += lpSum([coeff*x for coeff,x in zip(astringency,x)]) <= 4*lpSum(x), 'Upper Astringency constraint'
    
    problem += lpSum([coeff*x for coeff,x in zip(color,x)]) >= 4.5*lpSum(x), 'Lower Color constraint'
    problem += lpSum([coeff*x for coeff,x in zip(color,x)]) <= 5.5*lpSum(x), 'Upper Color constraint'
       
    # executing the optimization
    problem.solve()
    
    Qty = {}
    for var in problem.variables():
        Qty[var.name] = var.varValue
    
    return Qty, value(problem.objective)

def subtract_dict(d1,d2):
    d = {}
    for key in d1.keys():
        d[key] = d1[key] - d2[key]

    return d

D = {
    'January' : 600,
    'February' : 600,
    'March' : 700
}

# this is our starting values of the supply for the first month -> 'Jan'
# I will be updating these values after each month 
supplyCapacity = dict(zip(df.loc[:,'Region'], df.loc[:,'Qty Available (1,000 Gallons)']))

pQty = {}
cost = {}

for key in D.keys():
    pQty[key], cost[key] = optimize_single_month(df, D, key, list(supplyCapacity.values()))
    supplyCapacity = subtract_dict(supplyCapacity,pQty[key])

pd.set_option('precision',1)
    
summary = pd.DataFrame.from_dict(pQty,orient='columns')
summary.loc[:,'Purchased'] = summary.sum(axis=1)
summary.loc[:,'Available'] = df.set_index('Region').sort_index().loc[:,'Qty Available (1,000 Gallons)']

summary

January  February  March  Purchased  Available
Arizona         0.0      47.5   36.5       84.0         84
Brazil          0.0       0.0    0.0        0.0        672
California      0.0       0.0  111.2      111.2        168
China           0.0       0.0    0.0        0.0        588
Egypt         109.7      61.5   38.7      210.0        210
Florida       240.0     240.0  280.0      760.0       1200
India           0.0       0.0   13.5       13.5        400
Italy         108.2      71.8    0.0      180.0        180
Mexico        126.3     129.0   44.7      300.0        300
Spain           0.0      50.2  117.8      168.0        168
Texas          15.8       0.0   57.5       73.3        210